### PR TITLE
Respect Straight Join in Vitess query planning

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR adds the feature requested in https://github.com/vitessio/vitess/issues/15527.

After these changes, Vitess respects the order of tables specified in queries using straight joins.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/15527

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/onecase.json [hunk 1]
```diff
@@ -1,9 +1,60 @@
...
+  "OperatorType": "Join",
+  "Variant": "Join",
+  "JoinColumnIndexes": "R:0,R:1,L:0",
+  "JoinVars": {
+    "t1_id1": 1
+  },
+  "TableName": "t1_tbl",
+  "Inputs": [
+    {
+      "OperatorType": "Route",
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/onecase.json:11]: need to remove this on from onecase

File: go/vt/vtgate/planbuilder/operators/join.go [hunk 2]
```diff
@@ -61,7 +63,7 @@ func (j *Join) SetInputs(ops []Operator) {
 }
 
 func (j *Join) Compact(ctx *plancontext.PlanningContext) (Operator, *ApplyResult) {
-	if j.LeftJoin {
+	if !j.JoinType.IsCommutative() {
 		// we can't merge outer joins into a single QG
 		return j, NoRewrite
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/join.go:68]: comment not aligning with the condition check

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/join.go:68]: 2nd question should we not be allowed to merge keeping the original order for straight join

File: go/vt/vtgate/planbuilder/operators/ast_to_op.go [hunk 3]
```diff
@@ -228,8 +228,8 @@ func getOperatorFromJoinTableExpr(ctx *plancontext.PlanningContext, tableExpr *s
 	switch tableExpr.Join {
 	case sqlparser.NormalJoinType:
 		return createInnerJoin(ctx, tableExpr, lhs, rhs)
-	case sqlparser.LeftJoinType, sqlparser.RightJoinType:
-		return createOuterJoin(tableExpr, lhs, rhs)
+	case sqlparser.LeftJoinType, sqlparser.RightJoinType, sqlparser.StraightJoinType:
+		return createLeftAndStraightJoin(tableExpr, lhs, rhs)
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/ast_to_op.go:232]: if we use the same method of `left join` the inner join query for `straight join` will start disallowing `subquery` in the join condition

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 4]
```diff
@@ -139,11 +142,14 @@ func (aj *ApplyJoin) SetRHS(operator Operator) {
 }
 
 func (aj *ApplyJoin) MakeInner() {
-	aj.LeftJoin = false
+	if aj.IsInner() {
+		panic(vterrors.VT13001("Convert an already inner join"))
+	}
+	aj.JoinType = sqlparser.NormalJoinType
 }
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:149]: earlier this method was idempotent. Any reason to `panic` now on multiple calls of this method?

**Review by systay — APPROVED** [review]

File: go/test/endtoend/vtgate/gen4/gen4_test.go [hunk 5]
```diff
@@ -187,26 +187,6 @@ func TestSubQueriesOnOuterJoinOnCondition(t *testing.T) {
 	}
 }
 
-func TestPlannerWarning(t *testing.T) {
```
Comment (systay) [inline go/test/endtoend/vtgate/gen4/gen4_test.go:190]: Unfortunately, `straight_join` no longer produces a warning, and we don't have reliable ways of getting warnings from the planner without turning of schema tracking

File: go/vt/vtgate/planbuilder/operators/join.go [hunk 6]
```diff
@@ -83,38 +85,49 @@ func (j *Join) Compact(ctx *plancontext.PlanningContext) (Operator, *ApplyResult
...
-	if lok && rok {
-		op := &QueryGraph{
-			Tables:     append(lqg.Tables, rqg.Tables...),
-			innerJoins: append(lqg.innerJoins, rqg.innerJoins...),
-			NoDeps:     ctx.SemTable.AndExpressions(lqg.NoDeps, rqg.NoDeps),
-		}
-		return op
-	}
-	return &Join{LHS: LHS, RHS: RHS}
+	return op
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/join.go:116]: we should split the method into two as outer join and Straight Join have different code paths.

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**